src/shared/guards/roles.guard.ts

```diff
@@ -0,0 +1,34 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  SetMetadata,
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { Roles } from '../constants/user-roles.constants';
+
+const ROLES_DECORATOR_KEY = 'roles';
+@Injectable()
+export class RolesGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
+
+  canActivate(context: ExecutionContext) {
+    const allowedRoles = this.reflector.getAllAndOverride<Roles[]>(
+      ROLES_DECORATOR_KEY,
+      [context.getHandler(), context.getClass()],
+    );
+
+    if (!allowedRoles) {
+      return true;
+    }
+
+    const request = context.switchToHttp().getRequest();
+    const user = request.user;
+
+    return allowedRoles.includes(user.role);
```

Comment:
Is there an error to be thrown if false is returned? If not, throw an error here instead.

Reply:
yes it throws an error